cloudapi/v1/cloudapi.proto

```diff
@@ -6,49 +6,142 @@ import "google/protobuf/timestamp.proto";
+  INSTANCE_STATUS_OK = 2;
+  INSTANCE_STATUS_FAILED = 3;
+}
+
+message InstanceState {
+  InstanceStatus instance_status = 1;
+  optional string reason = 2;
+}
+
+service InstanceService {
+  rpc ListInstances(ListInstancesRequest) returns (ListInstancesResponse) {}
+  rpc CreateInstance(CreateInstanceRequest) returns (CreateInstanceResponse) {}
+  rpc UpdateInstance(UpdateInstanceRequest) returns (UpdateInstanceResponse) {}
+  rpc GetInstance(GetInstanceRequest) returns (GetInstanceResponse) {}
+  rpc DeleteInstance(DeleteInstanceRequest) returns (DeleteInstanceResponse) {}
+}
+
+// ListInstances via specific criteria, return all if no criteria specified
+message ListInstancesRequest {
+  optional InstanceFilter instance_filter = 1;
+}
+
+message InstanceFilter {
+  optional string name = 1;
+  optional string environment = 2;
+  // instance status from InstanceStatus enum
+  optional string status = 3;
+  optional string version = 4;
+  optional string hostname = 5;
+  optional string admin_email = 6;
+}
 
 message ListInstancesResponse {
   repeated Instance instances = 1;
   InstanceDrift drift = 2;
 }
 
-message Instance {
-  string id = 1;
-  string project_name = 2;
-  string type = 3;
-  oneof repo_details {
-    V1RepoDetails v1_repo_details = 4;
-    V2RepoDetails v2_repo_details = 5;
-  }
-  ProjectDetails project_details = 6;
-}
-
+// TODO move to instance details?
+// InstanceDrift returns informations about drift between repository and control-plane
 message InstanceDrift {
   repeated Instance only_repo = 1;
   repeated Instance only_project = 2;
 }
 
-message ProjectDetails {
-  string project_id = 1;
-  string email_domain = 2;
-  string instance_type = 3;
-  bool is_deleted = 4;
-  google.protobuf.Timestamp created_at = 5;
+// create instance
+message CreateInstanceRequest {
+  string name = 1;
+  string type = 2;
+  optional string environment = 3;
+  optional string version = 4;
+  optional string admin_email = 5;
+  optional string hostname = 6;
+  optional string license_key = 7;
+  optional string gcp_region = 8;
+  repeated string tags = 9;
 }
 
-message V1RepoDetails {
-  string customer = 1;
-  string email_domain = 2;
-  string endpoint = 3;
-  string instance_type = 4;
-  string type = 5;
-  bool disable_managed_smtp = 6;
-  bool disable_sourcegraph_management_access = 7;
+message CreateInstanceResponse {
+  Instance instance = 1;
+  optional string error = 2;
 }
 
-message V2RepoDetails {}
+/**
+   patch instance
+   https://google.aip.dev/134#patch-and-put
+   **/
+message UpdateInstanceRequest {
+  Instance instance = 1;
+}
 
-service InstanceService {
-  rpc ListInstances(ListInstancesRequest) returns (ListInstancesResponse) {}
+message UpdateInstanceResponse {
+  Instance instance = 1;
+  optional string error = 2;
+}
+
+message InstanceNameAndEnvironmentQuery {
+  string name = 1;
+  string environment = 2;
+}
+
+// get instance by any field
+message GetInstanceRequest {
+  oneof identifier {
```

Comment:
@michaellzc how about using oneof to ensure specific subset to identify given instance? for filter all are optional b/c there we use AND and can be more than 1 instance returned.

Reply:
how about for `GetInstanceRequest` we only support using global unique identifier to retrieve the instance? otherwise, they should use `List` with filters